Source/Dafny/Compilers/Compiler-go.cs

```diff
@@ -1920,35 +1921,62 @@ private class SpecialNativeType : UserDefinedType {
       var wBody = wr.NewBlock(");;");
       wBody.WriteLine("{0}, {1} := {2}()", tmpVarName, okVar, iterVar);
       wBody.WriteLine("if !{0} {{ break }}", okVar);
+      return wBody;
+    }
 
-      if (introduceBoundVar) {
-        wBody.WriteLine("var {0} {1}", boundVarName, TypeName(boundVarType, wBody, tok));
+    [CanBeNull]
+    protected override string GetSubtypeCondition(string tmpVarName, Type boundVarType, Bpl.IToken tok, ConcreteSyntaxTree wPreconditions) {
+      var conditions = new List<string> { };
+      if (boundVarType.IsNonNullRefType) {
+        conditions.Add($"!_dafny.IsDafnyNull({tmpVarName})");
       }
+
       if (boundVarType.IsRefType) {
-        var wIf = EmitIf($"_dafny.IsDafnyNull({tmpVarName})", true, wBody);
-        if (boundVarType.IsNonNullRefType) {
-          wIf.WriteLine("continue");
-        } else {
-          wIf.WriteLine("{0} = ({1})(nil)", boundVarName, TypeName(boundVarType, wBody, tok));
-        }
-        wIf = wBody.NewBlock("", open: BlockStyle.Brace);
-        string typeTest;
         if (boundVarType.IsObject || boundVarType.IsObjectQ) {
-          // nothing more to test
-          wIf.WriteLine("{0} = {1}.({2})", boundVarName, tmpVarName, TypeName(boundVarType, wIf, tok));
+          // Nothing more to test
         } else if (boundVarType.IsTraitType) {
           var trait = boundVarType.AsTraitType;
-          wIf.WriteLine($"if !_dafny.InstanceOfTrait({tmpVarName}.(_dafny.TraitOffspring), {TypeName_Companion(trait, wBody, tok)}.TraitID_) {{ continue }}");
-          wIf.WriteLine("{0} = {1}.({2})", boundVarName, tmpVarName, TypeName(boundVarType, wIf, tok));
+          conditions.Add(
+            $"_dafny.InstanceOfTrait/*1*/({tmpVarName}.(_dafny.TraitOffspring), {TypeName_Companion(trait, wPreconditions, tok)}.TraitID_)");
         } else {
-          typeTest = $"_dafny.InstanceOf({tmpVarName}, ({TypeName(boundVarType, wBody, tok)})(nil))";
-          wIf.WriteLine("{0}, {1} = {2}.({3})", boundVarName, okVar, tmpVarName, TypeName(boundVarType, wIf, tok));
-          wIf.WriteLine("if !{0} {{ continue }}", okVar);
+          var typeAssertSucceeds = idGenerator.FreshId("_typeAssertSucceeds");
+          wPreconditions.WriteLine(
+            $@"{typeAssertSucceeds} := func(param interface{{}}) bool {{ var ok bool; _, ok = param.({TypeName(boundVarType, wPreconditions, tok)}); return ok}}");
+          conditions.Add($"{typeAssertSucceeds}({tmpVarName})");
+          // When Go will support generics
```

Comment:
I'd leave out comments such as "change X when Y happens". I don't expect they'll make change X easier when we get to Y because we'll have forgotten about the comment, and maybe the comment never gets found and then it's just bloat.